File: ex211.py
from collections import defaultdict

class WordDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        
        self.w = defaultdict(set)
        

        
    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: void
        """
        
        # Add word to the set for its specific length
        self.w[len(word)].add(word)
        
        

    def search(self, word):
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """
            
        # If the word is exactly the same as the one in the set
        if word in self.w[len(word)]:
            return True
        
        # If the word contains "."
        if "." in word:
            for l in self.w[len(word)]:
                sign = 0
                
                for i in range(len(l)):
                    if (word[i] != '.') and (word[i] != l[i]):
                        sign = 1
                        break
                    
                # If every digit in word meet item in the set, True
                if sign == 0:
                    return True
                    
        return False
                        
                    
        
        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: code/ex211.cpp
class WordDictionary {
    struct TrieNode {
        vector<TrieNode*> children;
        bool ifEnd = 0;
        TrieNode() {
            children = vector<TrieNode*>(26);
        }
    };
    
public:
    TrieNode* root;
    
    /** Initialize your data structure here. */
    WordDictionary() {
        root = new TrieNode();
    }
    
    /** Adds a word into the data structure. */
    void addWord(string word) {
        TrieNode* p = root;
        for (char c : word) {
            if (p->children[c - 'a'] == NULL)
                p->children[c - 'a'] = new TrieNode();
            p = p->children[c - 'a'];
        }
        p->ifEnd = 1;
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    bool search(string word) {
        
        return helper(word, 0, root);
    }
    
    bool helper(string word, int s, TrieNode* p) {
        if (s >= word.length())    return p != NULL && p->ifEnd;
        for (int i = s; i < word.length(); i++) {
            char c = word[i];
            if (c != '.') {
                if (p == NULL)    return 0;
                if (p->children[c - 'a'] == NULL)    return 0;
                p = p->children[c - 'a'];
                if (i == word.size() - 1)    return p->ifEnd;
            } else {
                for (int j = 0; j < 26; j++) {
                    if (p->children[j] == NULL)    continue;
                    if (helper(word, i + 1, p->children[j]))    return 1;
                }
                return 0;
            }
        }
        return 0;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */